### ISSUE TITLE
update f4jumble to published crate

### ISSUE THREAD

**Issue by conorsch** [penumbra-zone/penumbra#4921]
The `penumbra-keys` crate depends on `f4jumble` via the zcash repo as a git dependency: https://github.com/penumbra-zone/penumbra/blob/v0.80.8/crates/core/keys/Cargo.toml#L30

In order to publish the workspace crates to crates.io (e.g. #3010), we can [only depend on other crates on crates.io](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-other-registries):

> crates.io does not allow packages to be published with dependencies on code published outside of crates.io.

Therefore we should switch to the most recent version of `f4jumble` published to crates.io, which is `v0.1.0`: https://crates.io/crates/f4jumble/0.1.0

**Comment by conorsch** [comment]
We want to upgrade `f4jumble` to `0.1.0`. To compare changes in the f4jumble source, we must compare the git rev version `2425a0869098e3b0588ccd73c42716bcf418612c` with that of the version published to crates.io as `0.1.0`. Here's a script I used to perform that comparison.

<details>
<summary>
bash script for comparing crate versions
</summary>

```
#!/bin/bash
# Quick script to compare two versions of the `f4jumble` dep,
# one from a specific git revision, the other from crates.io.
# Context for comparison is in
# https://github.com/penumbra-zone/penumbra/issues/4921
# Requires shell programs:
#
#   * curl
#   * diffoscope
#   * rsync
#   * tar
#
set -euo pipefail


# Create a local directory for unpacking the two archives,
# for comparison.
d="./workbench"
rm -rf "$d"
mkdir -p "$d"
cd "$d" || exit 1

# obtain the git rev version of the dep
git clone https://github.com/zcash/librustzcash
git -C librustzcash checkout 2425a0869098e3b0588ccd73c42716bcf418612c

parent_dir_for_git_rev_crate="f4jumble-0.1.0-git"
# structure the tar archive to have the same par dir name
# as the cargo version, to ease comparison.
parent_dir_for_git_rev_crate="f4jumble-0.1.0"
mkdir -p "$parent_dir_for_git_rev_crate"
rsync -a librustzcash/components/f4jumble/ "${parent_dir_for_git_rev_crate}/"
tar -cf f4jumble-git.tar "$parent_dir_for_git_rev_crate"
rm -r "$parent_dir_for_git_rev_crate"

# obtain the crates version:
# see here for discussion of this api call
# https://users.rust-lang.org/t/easiest-way-to-manually-download-a-crate-from-crates-io/67338
curl -sSfL https://crates.io/api/v1/crates/f4jumble/0.1.0/download | gunzip > f4jumble-cargo.tar

# compare the two, ignoring timestamps on the dirs, because we care only
# about differences in source code.
diffoscope --exclude-directory-metadata=recursive f4jumble-{git,cargo}.tar | tee f4jumble-diff.txt
```
</details>

Running that script locally, a substantive change to the `src/lib.rs` jumps out:

![f4jumble-diff](https://github.com/user-attachments/assets/97934c35-dd23-4803-9b17-af9808840343)

The return type has changed from an Option a Result, but otherwise looks the same. We can amend the very few call sites in our code to match. Elsewhere in the diff, I didn't see anything particularly problematic—it's mostly docs—but I'd very much like to have a +1 on that opinion.

<details>
<summary>
full diff of comparison
</summary>

```
--- f4jumble-git.tar
+++ f4jumble-cargo.tar
├── file list
│ @@ -1,9 +1,10 @@
│ -f4jumble-0.1.0/
│ -f4jumble-0.1.0/src/
│ -f4jumble-0.1.0/src/lib.rs
│ -f4jumble-0.1.0/src/test_vectors.rs
│ -f4jumble-0.1.0/src/test_vectors_long.rs
│ +f4jumble-0.1.0/.cargo_vcs_info.json
│ +f4jumble-0.1.0/CHANGELOG.md
│  f4jumble-0.1.0/Cargo.toml
│ +f4jumble-0.1.0/Cargo.toml.orig
│  f4jumble-0.1.0/LICENSE-APACHE
│  f4jumble-0.1.0/LICENSE-MIT
│  f4jumble-0.1.0/README.md
│ +f4jumble-0.1.0/src/lib.rs
│ +f4jumble-0.1.0/src/test_vectors.rs
│ +f4jumble-0.1.0/src/test_vectors_long.rs
├── f4jumble-0.1.0/src/lib.rs
│ @@ -1,8 +1,54 @@
│ +//! This crate provides a mechanism for "jumbling" byte slices in a reversible way.
│ +//!
│ +//! Many byte encodings such as [Base64] and [Bech32] do not have "cascading" behaviour:
│ +//! changing an input byte at one position has no effect on the encoding of bytes at
│ +//! distant positions. This can be a problem if users generally check the correctness of
│ +//! encoded strings by eye, as they will tend to only check the first and/or last few
│ +//! characters of the encoded string. In some situations (for example, a hardware device
│ +//! displaying on its screen an encoded string provided by an untrusted computer), it is
│ +//! potentially feasible for an adversary to change some internal portion of the encoded
│ +//! string in a way that is beneficial to them, without the user noticing.
│ +//!
│ +//! [Base64]: https://en.wikipedia.org/wiki/Base64
│ +//! [Bech32]: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#Bech32
│ +//!
│ +//! The function F4Jumble (and its inverse function, F4Jumble⁻¹) are length-preserving
│ +//! transformations can be used to trivially introduce cascading behaviour to existing
│ +//! encodings:
│ +//! - Prepare the raw `message` bytes as usual.
│ +//! - Pass `message` through [`f4jumble`] or [`f4jumble_mut`] to obtain the jumbled bytes.
│ +//! - Encode the jumbled bytes with the encoding scheme.
│ +//!
│ +//! Changing any byte of `message` will result in a completely different sequence of
│ +//! jumbled bytes. Specifically, F4Jumble uses an unkeyed 4-round Feistel construction to
│ +//! approximate a random permutation.
│ +//!
│ +//! ![Diagram of 4-round unkeyed Feistel construction](https://zips.z.cash/zip-0316-f4.png)
│ +//!
│ +//! ## Efficiency
│ +//!
│ +//! The cost is dominated by 4 BLAKE2b compressions for message lengths up to 128 bytes.
│ +//! For longer messages, the cost increases to 6 BLAKE2b compressions for 128 < lₘ ≤ 192,
│ +//! and 10 BLAKE2b compressions for 192 < lₘ ≤ 256, for example. The maximum cost for
│ +//! which the algorithm is defined would be 196608 BLAKE2b compressions at lₘ = 4194368.
│ +//!
│ +//! The implementations in this crate require memory of roughly lₘ bytes plus the size of
│ +//! a BLAKE2b hash state. It is possible to reduce this by (for example, with F4Jumble⁻¹)
│ +//! streaming the `d` part of the jumbled encoding three times from a less
│ +//! memory-constrained device. It is essential that the streamed value of `d` is the same
│ +//! on each pass, which can be verified using a Message Authentication Code (with key held
│ +//! only by the Consumer) or collision-resistant hash function. After the first pass of
│ +//! `d`, the implementation is able to compute `y`; after the second pass it is able to
│ +//! compute `a`; and the third allows it to compute and incrementally parse `b`. The
│ +//! maximum memory usage during this process would be 128 bytes plus two BLAKE2b hash
│ +//! states.
│ +
│  #![no_std]
│ +#![cfg_attr(docsrs, feature(doc_cfg))]
│  
│  use blake2b_simd::{Params as Blake2bParams, OUTBYTES};
│  
│  use core::cmp::min;
│  use core::fmt;
│  use core::ops::RangeInclusive;
│  use core::result::Result;
│ @@ -90,24 +136,24 @@
│          State { left, right }
│      }
│  
│      fn h_round(&mut self, i: u8) {
│          let hash = Blake2bParams::new()
│              .hash_length(self.left.len())
│              .personal(&H_PERS!(i))
│ -            .hash(&self.right);
│ +            .hash(self.right);
│          xor(self.left, hash.as_bytes())
│      }
│  
│      fn g_round(&mut self, i: u8) {
│          for j in 0..ceildiv(self.right.len(), OUTBYTES) {
│              let hash = Blake2bParams::new()
│                  .hash_length(OUTBYTES)
│                  .personal(&G_PERS!(i, j as u16))
│ -                .hash(&self.left);
│ +                .hash(self.left);
│              xor(&mut self.right[j * OUTBYTES..], hash.as_bytes());
│          }
│      }
│  
│      fn apply_f4jumble(&mut self) {
│          self.g_round(0);
│          self.h_round(0);
│ @@ -132,52 +178,128 @@
│      }
│  }
│  
│  fn ceildiv(num: usize, den: usize) -> usize {
│      (num + den - 1) / den
│  }
│  
│ +/// Encodes the given message in-place using F4Jumble.
│ +///
│ +/// Returns an error if the message is an invalid length. `message` will be unmodified in
│ +/// this case.
│ +///
│ +/// # Examples
│ +///
│ +/// ```
│ +/// let mut message_a = *b"The package from Alice arrives tomorrow morning.";
│ +/// f4jumble::f4jumble_mut(&mut message_a[..]).unwrap();
│ +/// assert_eq!(
│ +///     hex::encode(message_a),
│ +///     "861c51ee746b0313476967a3483e7e1ff77a2952a17d3ed9e0ab0f502e1179430322da9967b613545b1c36353046ca27",
│ +/// );
│ +///
│ +/// let mut message_b = *b"The package from Sarah arrives tomorrow morning.";
│ +/// f4jumble::f4jumble_mut(&mut message_b[..]).unwrap();
│ +/// assert_eq!(
│ +///     hex::encode(message_b),
│ +///     "af1d55f2695aea02440867bbbfae3b08e8da55b625de3fa91432ab7b2c0a7dff9033ee666db1513ba5761ef482919fb8",
│ +/// );
│ +/// ```
│  pub fn f4jumble_mut(message: &mut [u8]) -> Result<(), Error> {
│      if VALID_LENGTH.contains(&message.len()) {
│          State::new(message).apply_f4jumble();
│          Ok(())
│      } else {
│          Err(Error::InvalidLength)
│      }
│  }
│  
│ +/// Decodes the given message in-place using F4Jumble⁻¹.
│ +///
│ +/// Returns an error if the message is an invalid length. `message` will be unmodified in
│ +/// this case.
│ +///
│ +/// # Examples
│ +///
│ +/// ```
│ +/// let mut message_a = hex::decode(
│ +///     "861c51ee746b0313476967a3483e7e1ff77a2952a17d3ed9e0ab0f502e1179430322da9967b613545b1c36353046ca27")
│ +///     .unwrap();
│ +/// f4jumble::f4jumble_inv_mut(&mut message_a).unwrap();
│ +/// assert_eq!(message_a, b"The package from Alice arrives tomorrow morning.");
│ +///
│ +/// let mut message_b = hex::decode(
│ +///     "af1d55f2695aea02440867bbbfae3b08e8da55b625de3fa91432ab7b2c0a7dff9033ee666db1513ba5761ef482919fb8")
│ +///     .unwrap();
│ +/// f4jumble::f4jumble_inv_mut(&mut message_b).unwrap();
│ +/// assert_eq!(message_b, b"The package from Sarah arrives tomorrow morning.");
│ +/// ```
│  pub fn f4jumble_inv_mut(message: &mut [u8]) -> Result<(), Error> {
│      if VALID_LENGTH.contains(&message.len()) {
│          State::new(message).apply_f4jumble_inv();
│          Ok(())
│      } else {
│          Err(Error::InvalidLength)
│      }
│  }
│  
│ +/// Encodes the given message using F4Jumble, and returns the encoded message as a vector
│ +/// of bytes.
│ +///
│ +/// Returns an error if the message is an invalid length.
│ +///
│ +/// # Examples
│ +///
│ +/// ```
│ +/// let message_a = b"The package from Alice arrives tomorrow morning.";
│ +/// let encoded_a = f4jumble::f4jumble(message_a).unwrap();
│ +/// assert_eq!(
│ +///     hex::encode(encoded_a),
│ +///     "861c51ee746b0313476967a3483e7e1ff77a2952a17d3ed9e0ab0f502e1179430322da9967b613545b1c36353046ca27",
│ +/// );
│ +///
│ +/// let message_b = b"The package from Sarah arrives tomorrow morning.";
│ +/// let encoded_b = f4jumble::f4jumble(message_b).unwrap();
│ +/// assert_eq!(
│ +///     hex::encode(encoded_b),
│ +///     "af1d55f2695aea02440867bbbfae3b08e8da55b625de3fa91432ab7b2c0a7dff9033ee666db1513ba5761ef482919fb8",
│ +/// );
│ +/// ```
│  #[cfg(feature = "std")]
│ -pub fn f4jumble(message: &[u8]) -> Option<Vec<u8>> {
│ +#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
│ +pub fn f4jumble(message: &[u8]) -> Result<Vec<u8>, Error> {
│      let mut result = message.to_vec();
│ -    let res = f4jumble_mut(&mut result);
│ -    if res.is_ok() {
│ -        Some(result)
│ -    } else {
│ -        None
│ -    }
│ +    f4jumble_mut(&mut result).map(|()| result)
│  }
│  
│ +/// Decodes the given message using F4Jumble⁻¹, and returns the decoded message as a
│ +/// vector of bytes.
│ +///
│ +/// Returns an error if the message is an invalid length.
│ +///
│ +/// # Examples
│ +///
│ +/// ```
│ +/// let encoded_a = hex::decode(
│ +///     "861c51ee746b0313476967a3483e7e1ff77a2952a17d3ed9e0ab0f502e1179430322da9967b613545b1c36353046ca27")
│ +///     .unwrap();
│ +/// let message_a = f4jumble::f4jumble_inv(&encoded_a).unwrap();
│ +/// assert_eq!(message_a, b"The package from Alice arrives tomorrow morning.");
│ +///
│ +/// let encoded_b = hex::decode(
│ +///     "af1d55f2695aea02440867bbbfae3b08e8da55b625de3fa91432ab7b2c0a7dff9033ee666db1513ba5761ef482919fb8")
│ +///     .unwrap();
│ +/// let message_b = f4jumble::f4jumble_inv(&encoded_b).unwrap();
│ +/// assert_eq!(message_b, b"The package from Sarah arrives tomorrow morning.");
│ +/// ```
│  #[cfg(feature = "std")]
│ -pub fn f4jumble_inv(message: &[u8]) -> Option<Vec<u8>> {
│ +#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
│ +pub fn f4jumble_inv(message: &[u8]) -> Result<Vec<u8>, Error> {
│      let mut result = message.to_vec();
│ -    let res = f4jumble_inv_mut(&mut result);
│ -    if res.is_ok() {
│ -        Some(result)
│ -    } else {
│ -        None
│ -    }
│ +    f4jumble_inv_mut(&mut result).map(|()| result)
│  }
│  
│  #[cfg(test)]
│  mod common_tests {
│      use super::{f4jumble_inv_mut, f4jumble_mut, test_vectors};
│  
│      #[test]
│ @@ -194,19 +316,19 @@
│      #[test]
│      fn f4jumble_check_vectors_mut() {
│          #[cfg(not(feature = "std"))]
│          let mut cache = [0u8; test_vectors::MAX_VECTOR_LENGTH];
│          #[cfg(feature = "std")]
│          let mut cache = vec![0u8; test_vectors::MAX_VECTOR_LENGTH];
│          for v in test_vectors::TEST_VECTORS {
│ -            let mut data = &mut cache[..v.normal.len()];
│ -            data.clone_from_slice(&v.normal);
│ -            f4jumble_mut(&mut data).unwrap();
│ +            let data = &mut cache[..v.normal.len()];
│ +            data.clone_from_slice(v.normal);
│ +            f4jumble_mut(data).unwrap();
│              assert_eq!(data, v.jumbled);
│ -            f4jumble_inv_mut(&mut data).unwrap();
│ +            f4jumble_inv_mut(data).unwrap();
│              assert_eq!(data, v.normal);
│          }
│      }
│  }
│  
│  #[cfg(feature = "std")]
│  #[cfg(test)]
│ @@ -242,17 +364,17 @@
│              prop_assert_eq!(msg, unjumbled, "Unjumbled message did not match original message.");
│          }
│      }
│  
│      #[test]
│      fn f4jumble_check_vectors() {
│          for v in test_vectors::TEST_VECTORS {
│ -            let jumbled = f4jumble(&v.normal).unwrap();
│ +            let jumbled = f4jumble(v.normal).unwrap();
│              assert_eq!(jumbled, v.jumbled);
│ -            let unjumbled = f4jumble_inv(&v.jumbled).unwrap();
│ +            let unjumbled = f4jumble_inv(v.jumbled).unwrap();
│              assert_eq!(unjumbled, v.normal);
│          }
│      }
│  
│      #[test]
│      fn f4jumble_check_vectors_long() {
│          for v in test_vectors_long::TEST_VECTORS {
│ ├── encoding
│ │ @@ -1 +1 @@
│ │ -us-ascii
│ │ +utf-8
├── f4jumble-0.1.0/Cargo.toml
│ @@ -1,24 +1,38 @@
│ +# THIS FILE IS AUTOMATICALLY GENERATED BY CARGO
│ +#
│ +# When uploading crates to the registry Cargo will automatically
│ +# "normalize" Cargo.toml files for maximal compatibility
│ +# with all versions of Cargo and also rewrite `path` dependencies
│ +# to registry (e.g., crates.io) dependencies.
│ +#
│ +# If you are reading this file be aware that the original Cargo.toml
│ +# will likely look very different (and much more reasonable).
│ +# See Cargo.toml.orig for the original contents.
│ +
│  [package]
│ +edition = "2018"
│  name = "f4jumble"
│ +version = "0.1.0"
│ +authors = ["Jack Grigg <jack@electriccoin.co>", "Kris Nuttycombe <kris@electriccoin.co>", "Daira Hopwood <daira@electriccoin.co>"]
│  description = "Implementation of Zcash's F4Jumble algorithm"
│ -version = "0.0.0"
│ -authors = [
│ -    "Jack Grigg <jack@electriccoin.co>",
│ -    "Kris Nuttycombe <kris@electriccoin.co>",
│ -    "Daira Hopwood <daira@electriccoin.co>",
│ -]
│  homepage = "https://github.com/zcash/librustzcash"
│ -repository = "https://github.com/zcash/librustzcash"
│  readme = "README.md"
│ +keywords = ["feistel"]
│ +categories = ["encoding"]
│  license = "MIT OR Apache-2.0"
│ -edition = "2018"
│ -
│ -[dependencies]
│ -blake2b_simd = { version = "1", default-features = false }
│ +repository = "https://github.com/zcash/librustzcash"
│ +[package.metadata.docs.rs]
│ +all-features = true
│ +rustdoc-args = ["--cfg", "docsrs"]
│ +[dependencies.blake2b_simd]
│ +version = "1"
│ +default-features = false
│ +[dev-dependencies.hex]
│ +version = "0.4"
│  
│ -[dev-dependencies]
│ -proptest = "1"
│ +[dev-dependencies.proptest]
│ +version = "1"
│  
│  [features]
│  default = ["std"]
│  std = ["blake2b_simd/std"]
```
</details>

**Comment by conorsch** [comment]
Resolved in #4926.
